### ISSUE TITLE
Move "general purpose" parts of cmus out into a library

### ISSUE THREAD

**Issue by mahkoh** [cmus/cmus#541]
This can very well be a static library so that there is no change once the binary is linked. This would make it much easier to write unit tests for things such as #540.

**Comment by mahkoh** [comment]
To be more precise: There are parts that rely on global state, such as the cache. Those must stay where they are. But the parts that don't, such as the locking stuff, can be moved out into a library.

**Comment by flyingmutant** [comment]
Good idea. I'd only prefer to do this after the release, so that we have a stable reference point before refactoring.

Speaking of a release, I think -rc0 can be done once I'll find the cause of #505.

**Comment by nefthy** [comment]
I stumbled across this library[1] for another project. I think its worth some consideration, if we do refactoring anyway. It provides abstract collection with a nice consistent interface, which would make some parts of the code less tedious and more readable. They even implement Red-Black trees in HashSet. It is quite small, so we could even bundle it, for distros that don't ship the library.

[1] https://github.com/srdja/Collections-C/

**Comment by mahkoh** [comment]
The license is LGPL3 which would be a problem.

**Comment by nefthy** [comment]
too bad. the API and implementation look nice and clean.
